rcl/src/rcl/parameter_client.c

```diff
@@ -0,0 +1,437 @@
+  rcl_wait_set_t * wait_set,
+  const rcl_parameter_client_t * parameter_client)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(wait_set, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(parameter_client, RCL_RET_INVALID_ARGUMENT);
+  rcl_ret_t ret;
+
+  ret = rcl_wait_set_add_client(wait_set, &parameter_client->impl->get_client);
+  if (ret != RCL_RET_OK) {
+    RCL_SET_ERROR_MSG("Failed to add get_parameters client to waitset!");
+    return ret;
+  }
+  ret = rcl_wait_set_add_client(wait_set, &parameter_client->impl->get_types_client);
+  if (ret != RCL_RET_OK) {
+    RCL_SET_ERROR_MSG("Failed to add get_parameter_types client to waitset!");
+    return ret;
+  }
+  ret = rcl_wait_set_add_client(wait_set, &parameter_client->impl->set_client);
+  if (ret != RCL_RET_OK) {
+    RCL_SET_ERROR_MSG("Failed to add set_parameters client to waitset!");
+    return ret;
+  }
+  ret = rcl_wait_set_add_client(wait_set, &parameter_client->impl->set_atomically_client);
+  if (ret != RCL_RET_OK) {
+    RCL_SET_ERROR_MSG("Failed to add set_parameters_atomically client to waitset!");
+    return ret;
+  }
+  ret = rcl_wait_set_add_client(wait_set, &parameter_client->impl->list_client);
+  if (ret != RCL_RET_OK) {
+    RCL_SET_ERROR_MSG("Failed to add list_parameters client to waitset!");
+    return ret;
+  }
+
+  ret = rcl_wait_set_add_subscription(wait_set, &parameter_client->impl->event_subscription);
+  if (ret != RCL_RET_OK) {
+    RCL_SET_ERROR_MSG("Failed to add parameter events subscription to waitset!");
+    return ret;
+  }
+
+  return ret;
+}
+
+rcl_ret_t
+rcl_parameter_client_get_pending_action(
+  const rcl_wait_set_t * wait_set,
+  const rcl_parameter_client_t * parameter_client,
+  rcl_param_action_t * action)
+{
+  RCL_CHECK_ARGUMENT_FOR_NULL(wait_set, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(parameter_client, RCL_RET_INVALID_ARGUMENT);
+  RCL_CHECK_ARGUMENT_FOR_NULL(action, RCL_RET_INVALID_ARGUMENT);
+  size_t i = 0;
+  size_t j = 0;
+
+  for (i = 0; i < wait_set->size_of_clients; ++i) {
+    for (j = 0; j < RCL_NUMBER_OF_PARAMETER_ACTIONS; ++j) {
+      rcl_client_t * client_ptr = NULL;
+      *action = j;
+      switch (j) {
+        case RCL_GET_PARAMETERS:
+          client_ptr = &parameter_client->impl->get_client;
+          break;
+        case RCL_GET_PARAMETER_TYPES:
+          client_ptr = &parameter_client->impl->get_types_client;
+          break;
+        case RCL_SET_PARAMETERS:
+          client_ptr = &parameter_client->impl->set_client;
+          break;
+        case RCL_SET_PARAMETERS_ATOMICALLY:
+          client_ptr = &parameter_client->impl->set_atomically_client;
+          break;
+        case RCL_LIST_PARAMETERS:
+          client_ptr = &parameter_client->impl->list_client;
+          break;
+        default:
+          return RCL_RET_ERROR;
+      }
+      if (client_ptr == wait_set->clients[i]) {
+        return RCL_RET_OK;
+      }
+    }
+    *action = RCL_PARAMETER_ACTION_UNKNOWN;
```

Comment:
I don't know if this is relevant ; using `actions` here can get confusing if we implement ROS1-like actions in the future in rcl.